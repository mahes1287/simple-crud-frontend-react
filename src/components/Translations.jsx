import React, { useEffect, useState } from "react";
import axios from "axios";
import { Link } from "react-router-dom";
import DeleteDialog from "./DeleteDialog";
import Loading from "./Loading";

export default function Translations() {
  const [translations, setTranslations] = useState();
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [message, setMessage] = useState("");



  useEffect(() => {
    async function getData() {
      setLoading(true);
      const { data, error, message } = await translationsDataLoader();
      setTranslations(data);
      setError(error);
      setMessage(message);
      setLoading(false);
    }

    getData();
  }, []);

  if (loading) {
    return <Loading />;
  }
  if (error) {
    return (
      <div>
        <h1 className="text-2xl text-center">Translations</h1>
        There is following error: {message}. Contact site administrator
      </div>
    );
  }
  return (
    <div>
      <h1 className="text-2xl text-center ">List of Translations</h1>
      <div className="space-y-3 p-6">
        {translations.length ? (
          translations.map(({ input, output, fromUser, id }) => (
            <div
              key={id}
              className="flex flex-col space-y-2 max-w-7xl ml-5 mx-auto  bg-orange-100  rounded-lg shadow-md hover:bg-orange-200"
            >
              <div className="text-xl ml-3 py-3 truncate">
                <span className="italic ">input: </span> {input}
              </div>
              <div className="text-xl ml-3 py-3 truncate">
                <span className="italic ">output: </span> {output}
              </div>
              <div className="flex sm:flex-col md:flex-row justify-center items-center">
                <Link
                  to={`/translations/${id}/`}
                  className="m-2 font-semibold bg-green-100 hover:bg-green-400 px-4 rounded-lg py-2"
                >
                  view
                </Link>
                {localStorage.getItem("token") && (
                  <>
                    {" "}
                    <Link
                      to={`/translations/${id}/update/`}
                      state={{ input: input, output: output, id: id }}
                      className="m-2 font-semibold bg-blue-100 hover:bg-blue-400 px-4 rounded-lg py-2"
                    >
                      edit
                    </Link>
                    <DeleteDialog id={id}/>{" "}
                  </>
                )}
              </div>
            </div>
          ))
        ) : (
          <div>Its lonely here... Add some data..</div>
        )}
      </div>
    </div>
  );
}

export async function translationsDataLoader() {
  let config = {
    method: "get",
    url: `${process.env.REACT_APP_BASE_API_URL}/api/translations/`,
  };

  try {
    const response = await axios(config);
    return { data: response.data, error: false, message: null };
  } catch (error) {
    return {
      data: null,
      error: true,
      message: "Something went wrong",
    };
  }
}
