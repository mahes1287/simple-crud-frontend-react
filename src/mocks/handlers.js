// src/mocks/handlers.js
import { rest } from "msw";
const baseURL = process.env.REACT_APP_BASE_API_URL;

let mockTranslationsData = [
  {
    id: 1,
    input: "input1",
    output: "ouput1",
    fromUser: "Maheshwaran Velusamy",
  },
  {
    id: 2,
    input: "input2",
    output: "eeeeeerkkkkkkkkkkkkkkkkkkkkkkkk",
    fromUser: "Maheshwaran Velusamy",
  },
  {
    id: 3,
    input: "www",
    output: "dfdffd",
    fromUser: "Maheshwaran Velusamy",
  },
  {
    id: 4,
    input: "ooooooooooooooooooo",
    output: "iiiiiiiiiiiiiiiiiiiiiiii",
    fromUser: "Maheshwaran Velusamy",
  },
];

export const translationsLoadingHandler = rest.get(
  `${baseURL}/api/translations/`,
  async (req, res, ctx) => {
    localStorage.setItem("token", "true");
    localStorage.setItem("displayName", "Maheshwaran Velusamy");

    return res(ctx.status(200), ctx.delay());
  }
);

export const translationsZeroDataHandler = rest.get(
  `${baseURL}/api/translations/`,
  async (req, res, ctx) => {
    localStorage.setItem("token", "true");
    localStorage.setItem("displayName", "Maheshwaran Velusamy");

    return res(ctx.status(200), ctx.body([]));
  }
);

export const translationsHandler = rest.get(
  `${baseURL}/api/translations/`,
  async (req, res, ctx) => {
    localStorage.setItem("token", "true");
    localStorage.setItem("displayName", "Maheshwaran Velusamy");

    return res(ctx.status(200), ctx.body(mockTranslationsData));
  }
);
export const translationsErrorHandler = rest.get(
  `${baseURL}/api/translations/`,
  async (req, res, ctx) => {
    return res(ctx.status(500), ctx.body());
  }
);

export const handlers = [translationsHandler];
