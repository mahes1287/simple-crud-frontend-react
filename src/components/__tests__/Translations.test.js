import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import {
  translationsErrorHandler,
  translationsHandler,
  translationsLoadingHandler,
  translationsZeroDataHandler,
} from "../../mocks/handlers";
import { server } from "../../mocks/server";
import Translations from "../Translations";

describe("<Translations />", () => {
  test("should render Loading screen", () => {
    server.use(translationsLoadingHandler);
    const { asFragment } = render(
      <MemoryRouter>
        <Translations />
      </MemoryRouter>
    );

    const loadingText = screen.getByRole("heading");
    expect(loadingText).toHaveTextContent("Loading...");
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render -Its lonely here...- when data from api is empty array []", async () => {
    server.use(translationsZeroDataHandler);
    const { asFragment } = render(
      <MemoryRouter>
        <Translations />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByText("List of Translations")).toBeInTheDocument();
    });
    const headingText = screen.getByText("List of Translations");
    expect(headingText).toHaveTextContent("List of Translations");
    expect(
      screen.getByText("Its lonely here... Add some data..")
    ).toHaveTextContent("Its lonely here... Add some data..");

    expect(asFragment()).toMatchSnapshot();
  });
  test("should render list  of translations when data available", async () => {
    server.use(translationsHandler);
    const { asFragment } = render(
      <MemoryRouter>
        <Translations />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(
        screen.getByText("eeeeeerkkkkkkkkkkkkkkkkkkkkkkkk")
      ).toBeInTheDocument();
    });
    const headingText = screen.getByText("List of Translations");
    expect(headingText).toHaveTextContent("List of Translations");
    expect(
      screen.getByText("eeeeeerkkkkkkkkkkkkkkkkkkkkkkkk")
    ).toHaveTextContent("eeeeeerkkkkkkkkkkkkkkkkkkkkkkkk");

    expect(asFragment()).toMatchSnapshot();
  });
  test("should render error while fetching data", async () => {
    server.use(translationsErrorHandler);
    const { asFragment } = render(
      <MemoryRouter>
        <Translations />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByText(/There is following error/i)).toBeInTheDocument();
    });

    expect(screen.getByText(/There is following error/i)).toHaveTextContent(
      /There is following error/i
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
