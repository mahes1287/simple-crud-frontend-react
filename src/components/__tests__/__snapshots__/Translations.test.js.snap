// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Translations /> should render -Its lonely here...- when data from api is empty array [] 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-2xl text-center "
    >
      List of Translations
    </h1>
    <div
      class="space-y-3 p-6"
    >
      <div>
        Its lonely here... Add some data..
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Translations /> should render Loading screen 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto text-center p-10 space-y-3"
  >
    <span
      style="display: inherit;"
    >
      <span
        style="background-color: rgb(54, 55, 214); width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-SyncLoader-sync 0.6s 0.07s infinite ease-in-out; animation-fill-mode: both;"
      />
      <span
        style="background-color: rgb(54, 55, 214); width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-SyncLoader-sync 0.6s 0.14s infinite ease-in-out; animation-fill-mode: both;"
      />
      <span
        style="background-color: rgb(54, 55, 214); width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-SyncLoader-sync 0.6s 0.21000000000000002s infinite ease-in-out; animation-fill-mode: both;"
      />
    </span>
    <h1>
      Loading...
    </h1>
  </div>
</DocumentFragment>
`;

exports[`<Translations /> should render error while fetching data 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-2xl text-center"
    >
      Translations
    </h1>
    There is following error: Something went wrong. Contact site administrator
  </div>
</DocumentFragment>
`;

exports[`<Translations /> should render list  of translations when data available 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-2xl text-center "
    >
      List of Translations
    </h1>
    <div
      class="space-y-3 p-6"
    >
      <div
        class="flex flex-col space-y-2 max-w-7xl ml-5 mx-auto  bg-orange-100  rounded-lg shadow-md hover:bg-orange-200"
      >
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            input: 
          </span>
           input1
        </div>
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            output: 
          </span>
           ouput1
        </div>
        <div
          class="flex sm:flex-col md:flex-row justify-center items-center"
        >
          <a
            class="m-2 font-semibold bg-green-100 hover:bg-green-400 px-4 rounded-lg py-2"
            href="/translations/1/"
          >
            view
          </a>
           
          <a
            class="m-2 font-semibold bg-blue-100 hover:bg-blue-400 px-4 rounded-lg py-2"
            href="/translations/1/update/"
          >
            edit
          </a>
          <button
            class="m-2 font-semibold bg-red-100 hover:bg-red-400 px-4 rounded-lg py-2"
            type="button"
          >
            Delete
          </button>
           
        </div>
      </div>
      <div
        class="flex flex-col space-y-2 max-w-7xl ml-5 mx-auto  bg-orange-100  rounded-lg shadow-md hover:bg-orange-200"
      >
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            input: 
          </span>
           input2
        </div>
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            output: 
          </span>
           eeeeeerkkkkkkkkkkkkkkkkkkkkkkkk
        </div>
        <div
          class="flex sm:flex-col md:flex-row justify-center items-center"
        >
          <a
            class="m-2 font-semibold bg-green-100 hover:bg-green-400 px-4 rounded-lg py-2"
            href="/translations/2/"
          >
            view
          </a>
           
          <a
            class="m-2 font-semibold bg-blue-100 hover:bg-blue-400 px-4 rounded-lg py-2"
            href="/translations/2/update/"
          >
            edit
          </a>
          <button
            class="m-2 font-semibold bg-red-100 hover:bg-red-400 px-4 rounded-lg py-2"
            type="button"
          >
            Delete
          </button>
           
        </div>
      </div>
      <div
        class="flex flex-col space-y-2 max-w-7xl ml-5 mx-auto  bg-orange-100  rounded-lg shadow-md hover:bg-orange-200"
      >
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            input: 
          </span>
           www
        </div>
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            output: 
          </span>
           dfdffd
        </div>
        <div
          class="flex sm:flex-col md:flex-row justify-center items-center"
        >
          <a
            class="m-2 font-semibold bg-green-100 hover:bg-green-400 px-4 rounded-lg py-2"
            href="/translations/3/"
          >
            view
          </a>
           
          <a
            class="m-2 font-semibold bg-blue-100 hover:bg-blue-400 px-4 rounded-lg py-2"
            href="/translations/3/update/"
          >
            edit
          </a>
          <button
            class="m-2 font-semibold bg-red-100 hover:bg-red-400 px-4 rounded-lg py-2"
            type="button"
          >
            Delete
          </button>
           
        </div>
      </div>
      <div
        class="flex flex-col space-y-2 max-w-7xl ml-5 mx-auto  bg-orange-100  rounded-lg shadow-md hover:bg-orange-200"
      >
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            input: 
          </span>
           ooooooooooooooooooo
        </div>
        <div
          class="text-xl ml-3 py-3 truncate"
        >
          <span
            class="italic "
          >
            output: 
          </span>
           iiiiiiiiiiiiiiiiiiiiiiii
        </div>
        <div
          class="flex sm:flex-col md:flex-row justify-center items-center"
        >
          <a
            class="m-2 font-semibold bg-green-100 hover:bg-green-400 px-4 rounded-lg py-2"
            href="/translations/4/"
          >
            view
          </a>
           
          <a
            class="m-2 font-semibold bg-blue-100 hover:bg-blue-400 px-4 rounded-lg py-2"
            href="/translations/4/update/"
          >
            edit
          </a>
          <button
            class="m-2 font-semibold bg-red-100 hover:bg-red-400 px-4 rounded-lg py-2"
            type="button"
          >
            Delete
          </button>
           
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
